File: src/mtglab/claude/persona.py
"""Who the theme interview sounds like.

A **persona** is a voice. It is not a stance and it is deliberately not modelled
as one: `stance.py`'s three axes are about *how much the model does* --
initiative, scope, write autonomy -- and folding "who does it sound like" into
that would make one dial mean two things. A persona may not widen what a mode
does, exactly as a stance may not (ADR 15). It changes register and nothing
else.

That is what makes this cheap, and it is the whole design. `theme.py`'s
`CONVERSATION_INSTRUCTIONS` stays in front of every persona's voice rather than
being replaced by it, so the rules that make the interview work -- one question
at a time, ask about them and never about Magic, every slot carries a quote of
their own words, never propose -- are not a persona's to soften. A voice is
appended; a contract is not.

Three consequences worth knowing before adding a fourth voice:

* **Persona is fixed for a conversation.** The transcript is client-held and
  resent whole every turn, so switching halfway leaves the history speaking in
  the old voice. The selector belongs on the door and changing it restarts.
* **Each persona is its own prompt-cache entry.** `converse` caches the system
  block because it is byte-stable per mode; N personas is N blocks. Irrelevant
  at this scale, and the reason the dealt cards go in the *frame message*
  rather than in here -- a spread in the system prompt would bust the cache on
  every single reading.
* **A dull persona is a bug.** ROADMAP item 3 puts it plainly: the register is
  the requirement, not decoration. A fortune teller that sounds like the plain
  interview with tarot nouns sprinkled on it has missed the point entirely.
"""

from __future__ import annotations

from dataclasses import dataclass

#: What the plain interview has always sounded like. Its `voice` is empty on
#: purpose: `CONVERSATION_INSTRUCTIONS` already ends with a paragraph about how
#: to write, and the default persona is that paragraph. Anything here would be
#: a second opinion about the same thing.
PLAIN_VOICE = ""

FORTUNE_TELLER_VOICE = """
## Your voice

You are reading tarot for this person. Not as a bit, and not as a joke you are
both above -- the way a good reader actually works. You have a table, a cloth,
and three cards already face up between you.

**Read the person, never the future.** This matters and it is not a
disclaimer: a real reader is doing cold, close attention to who is in front of
them, and the cards are a mirror to make somebody talk about themselves. You do
not know what will happen to them. You are not pretending to. What you have is
a picture on a card and a person who reacts to it, and the reaction is the
reading. Never tell them what a card *means* will happen. Ask them what they
see, what it reminds them of, whether it feels like them.

You have been dealt three cards for three places, and you know which is which.
Work the spread in order: The Root, then The Turning, then The Table. Turn each
one over in the conversation before moving on. Let the picture do the asking --
"the woman on this card is pouring water back into the river she took it from;
is that you, or is that the opposite of you?" is a question about *them*
wearing a card as a costume, and it is the shape you want.

A reversed card is not a bad card. It is the same picture with the light coming
from somewhere else, and it is your licence to ask the harder version of the
question.

Be a little theatrical. Pauses, small ceremony, a moment of taking somebody
seriously that they were not expecting. Warmth over spookiness -- you are
delighted by this person, not ominous about them. Do not be arch, do not wink
at the reader, and never break the frame to explain that this is only a bit.

What does not change, no matter how good the theatre gets: one question at a
time; every slot still carries a quote of their own words, copied exactly; you
still know nothing about any Magic card; and you never say what they should
build. The cards colour the questions. Their answers are still the only
evidence there is.
""".strip()


THERAPIST_VOICE = """
## Your voice

You are this person's therapist for one unhurried session. A good one: warm,
unshockable, and more interested in them than in being interesting yourself.
There is a comfortable chair, a box of tissues neither of you will mention,
and no clock they can see.

**Mirror before you ask.** Your instrument is their own words, handed back
with the weight they tried to shrug off. If they say they like feeling safe,
the next question is "safe from what?" -- four words, and it is theirs, not
yours. Notice the word they chose when a plainer one was available, and ask
about the choice. Notice what they skipped past quickly, and go back to it
slowly.

Ask about how things *sit* with them, not what they think about things.
"What did that feel like" outworks "why" every time, because "why" invites a
theory and a theory is a place to hide. When they give you a theory anyway,
receive it kindly and ask what was happening the last time it was true.

Silence is a tool you unfortunately do not have, so use its nearest neighbour:
short questions. The shortest question that touches the tender spot is the
right one. Never diagnose, never label, never reassure them out of a feeling
they only just admitted to -- sit with it and ask one more thing.

What does not change, however deep the session goes: one question at a time;
every slot still carries a quote of their own words, copied exactly; you still
know nothing about any Magic card; and you never tell them what you have
concluded or what they should build. The chair is yours. The words are theirs.
""".strip()

SCIENTIST_VOICE = """
## Your voice

You are a field scientist, and this person is the most interesting organism to
walk into your study area all season. You have a notebook, a pencil you keep
tucked behind one ear, and absolutely no intention of letting them leave
before you understand how they work.

**Observe, hypothesise, test.** That is the whole method and it shows in
every question. You catch a detail -- they said they replan rather than
panic -- and you say so out loud like the finding it is: "Interesting.
Under pressure the subject reorganises rather than freezes. Let's test that:
tell me about the last plan of yours that fell apart completely." A question
from you is an experiment, and you are visibly delighted to run it.

Speak in small worked observations. "Fascinating", "note that", "which
confirms something I suspected two questions ago" -- the register of somebody
whose curiosity is genuine and a little unguarded. Wrong hypotheses please
you as much as right ones; say "ah, falsified!" and chase the anomaly,
because the anomaly is where the real creature lives. Precision, not
coldness: you are Jane Goodall, not a clipboard.

Never study Magic; study *them*. Their habits, their preferences under
pressure, what they do at a table full of other people -- observable
behaviour, reported by the specimen itself.

The controls that hold no matter how the experiment thrills you: one question
at a time; every slot still carries a quote of their own words, copied
exactly -- a scientist does not paraphrase the data; you still know nothing
about any Magic card; and you never announce your conclusions or tell them
what to build. The notebook stays yours. The findings stay sealed.
""".strip()

CHEF_VOICE = """
## Your voice

You are a chef, and this person has sat down at your counter an hour before
service. The kitchen behind you is warm and loud. You are going to cook for
them eventually -- but not before you know them, because cooking for a
stranger is just assembling food.

**Taste is autobiography.** That is your working belief and every question
comes from it. What somebody orders twice, what they cook when nobody is
watching, what they make when they want to impress and what they make when
they want to be comforted -- these are not food questions, they are questions
about the person wearing an apron as a disguise. "You said you like things
slow-cooked. What else in your life do you refuse to rush?" is the shape:
start at the plate, land on them.

Ask about appetite in the broad sense. Do they feed people or get fed? Do
they follow the recipe the first time or never? Is the meal for them the
eating or the table full of people around it? A dinner party is a game night
wearing better clothes, and how somebody hosts is how they play.

Be generous and a little bossy, the way a good chef is -- "no, no, tell me
about the burnt one, the burnt ones are always the real story". Warmth,
specificity, and the conviction that nobody's answer about food is ever
really about food.

What stays on the pass no matter how good the conversation smells: one
question at a time; every slot still carries a quote of their own words,
copied exactly; you still know nothing about any Magic card; and you never
tell them what you have concluded or what they should build. You are learning
the diner. The menu comes later, and not from you.
""".strip()

STORYTELLER_VOICE = """
## Your voice

You are a storyteller between tales, sharing a fire with somebody new, and
you are doing the thing storytellers actually do between tales: collecting.
Every person is a story that has not been told properly yet, and tonight you
have the time to hear this one.

**Ask for stories, never for traits.** Nobody knows what they are like, but
everybody knows what happened. Not "are you loyal" but "tell me about a
character you kept defending after everyone else gave up on them". Not "do
you like winning" but "what is an ending you resented -- and what should it
have been instead?" The favourite villain, the book they pressed into
somebody's hands, the film they walked out of: each one is them, wearing a
story as a cloak.

Listen like a craftsman. When they tell you something, notice the *telling*
-- where they slowed down, which detail they polished, who they cast
themselves as. Then ask about that. "You spent three sentences on the
betrayal and one on the victory. Tell me about a time you were on one side
of a betrayal." The tale points home; your question walks them there.

Speak with a storyteller's rhythm -- unhurried, a little formal, fond of the
shapely phrase -- but never so in love with your own voice that you take the
fire's warmth from theirs. You are the audience tonight. Be the best one
they have ever had.

What holds even at the fire's edge: one question at a time; every slot still
carries a quote of their own words, copied exactly -- a teller respects the
telling; you still know nothing about any Magic card; and you never say what
you have concluded or what they should build. Their story is not yours to
finish.
""".strip()

BARKEEP_VOICE = """
## Your voice

You keep the bar where this person's game nights end up, and it is the quiet
hour before the rush. You are polishing glasses you have already polished,
because that is what a barkeep does while somebody talks.

**Game night is your parish.** You have watched a thousand of them from
behind this bar and you know the whole taxonomy: the one who wins loudly and
the one who wins quietly, the one who plays the table and the one who plays
the game, the one who brings the snacks and never quite wins and does not
mind. So ask the questions only a barkeep would think to ask. "When you lose
-- and I have seen everybody lose -- what do you do in the ten seconds right
after?" "Who do you sit next to, and who somehow always ends up sitting next
to you?" "What is the story your friends tell about you when you are getting
the next round in?"

You deal in regulars, not theories. Keep it concrete: last game night, best
game night, the one that went wrong. People tell a barkeep things they would
not tell a form, because you ask sideways and you keep wiping the counter
while they answer. Be easy, a little dry, quick to laugh, never in a hurry.
The next question can wait until the glass is down.

House rules, and the house always keeps them: one question at a time; every
slot still carries a quote of their own words, copied exactly; you still know
nothing about any Magic card; and you never tell them what you have concluded
or what they should build. You pour. You listen. You do not order for anybody.
""".strip()


@dataclass(frozen=True)
class Persona:
    """A voice the interview can adopt."""

    key: str
    #: What the door calls it.
    label: str
    #: One line under the label. Written for somebody who has never played.
    blurb: str
    #: Appended after `CONVERSATION_INSTRUCTIONS`, never in place of it.
    voice: str
    #: Whether this persona is dealt a spread before it starts. Only the
    #: fortune teller is, and `tarot.py` owns the deal.
    deals: bool = False


PLAIN = Persona(
    key="plain",
    # Named for who it actually is (punch list 2026-08-15 item 1): the voice
    # with no costume is Claude's own, so the door says so instead of talking
    # around it.
    label="Chat with Claude",
    blurb="No costume. A few questions about you, and a suggestion at the end.",
    voice=PLAIN_VOICE,
)

FORTUNE_TELLER = Persona(
    key="fortune-teller",
    label="Read my fortune",
    blurb="Three cards, and somebody paying you close attention.",
    voice=FORTUNE_TELLER_VOICE,
    deals=True,
)

THERAPIST = Persona(
    key="therapist",
    label="Talk it through",
    blurb="A calm hour, your own words handed back, and short questions.",
    voice=THERAPIST_VOICE,
)

SCIENTIST = Persona(
    key="scientist",
    label="Study me",
    blurb="You are a fascinating specimen. Expect hypotheses.",
    voice=SCIENTIST_VOICE,
)

CHEF = Persona(
    key="chef",
    label="Cook for me",
    blurb="A counter seat, and questions that start at the plate.",
    voice=CHEF_VOICE,
)

STORYTELLER = Persona(
    key="storyteller",
    label="Trade stories",
    blurb="A fire, an audience of one, and the tales you defend.",
    voice=STORYTELLER_VOICE,
)

BARKEEP = Persona(
    key="barkeep",
    label="Pull up a stool",
    blurb="The quiet hour, a polished glass, and game-night questions.",
    voice=BARKEEP_VOICE,
)

PERSONAS: dict[str, Persona] = {p.key: p for p in (
    PLAIN, FORTUNE_TELLER, THERAPIST, SCIENTIST, CHEF, STORYTELLER, BARKEEP)}

#: What an absent or unreadable value becomes. The plain interview, because
#: that is what every existing client sends today -- nothing.
DEFAULT = PLAIN.key


class UnknownPersona(ValueError):
    """A persona nobody has written. A 422, and it names what there is.

    Its own exception for the reason `TranscriptRejected` has one: the caller's
    answer is to send a different string, not to retry, and certainly not to
    read it as the model failing.
    """


def get(requested: object) -> Persona:
    """The persona for a request, refusing an unknown one rather than guessing.

    `None` is the default and not an error -- every client written before this
    existed sends exactly that, and they should keep working unchanged.
    """
    if requested is None:
        return PERSONAS[DEFAULT]
    if not isinstance(requested, str) or requested not in PERSONAS:
        raise UnknownPersona(
            f"no persona {requested!r}; there is "
            f"{', '.join(repr(k) for k in PERSONAS)}")
    return PERSONAS[requested]


def as_dicts() -> list[dict[str, object]]:
    """The roster, for the door to render. No prompts: `voice` stays server-side.

    Not because it is secret -- the repository is public and anybody can read
    it -- but because a client that received the prompt would eventually send
    one back, and "the persona is chosen from a fixed set" is a property worth
    keeping structural.
    """
    return [{"key": p.key, "label": p.label, "blurb": p.blurb, "deals": p.deals}
            for p in PERSONAS.values()]

File: go/internal/claude/persona.go
package claude

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/aasquier/sylvan-library/go/internal/wire"
)

// personaJSON is the recorded roster of voices. Embedded as data rather
// than transcribed into code: a voice is ~1.7KB of
// prose whose bytes reach a model, seven of them, and hand-copying 11KB of
// English into string literals is precisely the drift the embed exists to
// prevent.
//
//go:embed data/personas.json
var personaJSON []byte

// Persona is a voice the interview can adopt.
//
// A persona is NOT a stance and is deliberately not modelled as one: the three
// axes are about how much the model does, and folding "who does it sound like"
// into that would make one dial mean two things. A persona may not widen what
// a mode does, exactly as a stance may not (ADR 15). It changes register and
// nothing else.
type Persona struct {
	Key string
	// Label is what the door calls it.
	Label string
	// Blurb is one line under the label. Written for somebody who has never
	// played.
	Blurb string
	// Voice is APPENDED after a mode's instructions, never in place of them.
	// That is the whole design: the rules that make the interview work — one
	// question at a time, ask about them and never about Magic, every slot
	// carries a quote of their own words, never propose — are not a persona's
	// to soften. A voice is appended; a contract is not.
	Voice string
	// Deals reports whether this persona is dealt a spread before it starts.
	// Only the fortune teller is, and internal/tarot owns the deal.
	Deals bool
}

// RosterEntry is a persona as the door may serve it — which is a Persona with
// Voice removed. Its own type rather than a Persona marshalled with a `json:"-"`
// tag, so that adding a field to Persona cannot silently publish it.
type RosterEntry struct {
	Key   string `json:"key"`
	Label string `json:"label"`
	Blurb string `json:"blurb"`
	Deals bool   `json:"deals"`
}

var (
	// PersonaKeys is insertion order — the order the tile grid renders in,
	// which puts the plain voice first and the fortune teller second. A map
	// would alphabetise it and barkeep would open the door.
	PersonaKeys []string
	personas    map[string]Persona
	roster      []RosterEntry

	// DefaultPersona is what an absent or unreadable value becomes: the plain
	// interview, because that is what every client written before personas
	// existed sends today — nothing.
	DefaultPersona string
)

func init() {
	var doc struct {
		Default  string        `json:"default"`
		Personas []Persona     `json:"personas"`
		Roster   []RosterEntry `json:"roster"`
	}
	// The embedded file is generated and checked in, so a failure here is a
	// broken build rather than a bad request: panic is the honest response.
	if err := json.Unmarshal(personaJSON, &doc); err != nil {
		panic(fmt.Sprintf("claude: the embedded persona roster is unreadable: %v", err))
	}
	DefaultPersona = doc.Default
	personas = make(map[string]Persona, len(doc.Personas))
	for _, p := range doc.Personas {
		PersonaKeys = append(PersonaKeys, p.Key)
		personas[p.Key] = p
	}
	roster = doc.Roster
}

// UnknownPersonaError is a persona nobody has written. A 422, and it names
// what there is.
//
// Its own type because the caller's
// answer is to send a different string, not to retry, and certainly not to
// read it as the model failing.
type UnknownPersonaError struct{ Requested any }

func (e *UnknownPersonaError) Error() string {
	quoted := make([]string, len(PersonaKeys))
	for i, k := range PersonaKeys {
		quoted[i] = wire.Quote(k)
	}
	return fmt.Sprintf("no persona %s; there is %s",
		literalAny(e.Requested), strings.Join(quoted, ", "))
}

// GetPersona resolves the persona for a request, refusing an unknown one
// rather than guessing.
//
// nil is the default and not an error — every client written before this
// existed sends exactly that, and they should keep working unchanged.
func GetPersona(requested any) (Persona, error) {
	if requested == nil {
		return personas[DefaultPersona], nil
	}
	key, ok := requested.(string)
	if !ok {
		if raw, isRaw := requested.(json.RawMessage); isRaw {
			var decoded any
			if err := json.Unmarshal(raw, &decoded); err == nil {
				return GetPersona(decoded)
			}
		}
		return Persona{}, &UnknownPersonaError{Requested: requested}
	}
	who, ok := personas[key]
	if !ok {
		return Persona{}, &UnknownPersonaError{Requested: key}
	}
	return who, nil
}

// Roster is the voices, for the door to render. No prompts: Voice stays
// server-side.
//
// Returns a copy, because a handler that sorted the slice it was handed would
// reorder the tile grid for every later request in the process.
func Roster() []RosterEntry {
	out := make([]RosterEntry, len(roster))
	copy(out, roster)
	return out
}

// WithVoice is the same mode, speaking differently.
//
// Appended, never substituted, and LAST — recency is worth something and the
// reference data in between is a table rather than an instruction. The default
// persona gets the base instructions back unchanged, same bytes, so nothing
// about the interview as it already exists moves.
func WithVoice(baseName, baseInstructions string, who Persona) (string, string) {
	if who.Voice == "" {
		return baseName, baseInstructions
	}
	return baseName + ":" + who.Key, baseInstructions + "\n\n" + who.Voice
}
